### ISSUE TITLE
[Question] Is there a way to get custom WsFed parameters mapped over to SignInMessage for a custom UserService?

### ISSUE THREAD

**Issue by rayph** [IdentityServer/Thinktecture.IdentityServer3.WsFederation#19]
I have regular authentication working nicely, but I'd like to build out a slightly special journey where folks can automatically sign in as a particular user, "guestX" or "guestY" for example.

The best way seems to be to get parameter(s) into a custom UserService where an override of PreAuthenticateAsync(SignInMessage) could then decide whether to return an AuthenticationResult. 

I have tweaked the existing code in WsFederationController to simply add a Login Hint just before the cookie gets written but was wondering if something more out of the box could be dynamically mapped from SignInRequestMessage(WsFed params) to the SignInMessage. 

To summarise; I am hoping to put values into AuthenticationProperties at the point of initiating the sign in (on the RP side) which can then be evaluated by the UserService (IdSrv) which then *may* automatically sign the user in.

Anyone have thoughts on this?

**Comment by brockallen** [comment]
There are some other threads in the core repo that ask about this. Of course, they're using OIDC for the protocol. We don't have a sample for this, but the approach would be to use the acr_values to indicate you want this "delegation" and use PreAuthN like you suggest to get the user logged in without credentials. So I think you're on the right track.

**Comment by leastprivilege** [comment]
We don't support custom parameters for WS-Federation. We do for OIDC.

Check the ws-fed spec if you find a suitable parameter that could be used for extensibility.

**Comment by rayph** [comment]
I think I would like to make use of either the wfed or maybe the legacy wres parameter. Not sure how to get these passed through from the RP side yet (using Microsoft.Owin.Security.WsFederation)

wfed -> SignInRequestMessage.Federation (context id or any string?)
wres -> SignInRequestMessage.Resource (legacy/unused?)

Anything where a value could be passed all the way through would be useful...

**Comment by brockallen** [comment]
If you're going to update the WS-Fed RP to pass custom params, why can't you update it for OIDC?

**Comment by rayph** [comment]
Aiming to get full stack implementation working using WsFed for now, OIDC would be a bit further down the line 

Plugging in "wfed" as an option for RP by overriding WsFederationAuthenticationHandler was quite straightforward actually. Most things there are virtual.

**Comment by leastprivilege** [comment]
What kind of extra data do you want to pass to the wsfed endpoint?

**Comment by rayph** [comment]
I'm thinking of using an encrypted(url-safe) bit of data that the UserService.PreAuthenticateAsync can act on. It may automatically authenticate the user as a guest login for instance.

**Comment by rayph** [comment]
It would be pretty handy if "wfed" was forwarded by the WsFederationController maybe as LoginHint or perhaps an entirely new property on SignInMessage to the AuthenticationController. I imagine other folks might want to forward a parameter from the RP like this or am I doing completely unexpected?

**Comment by leastprivilege** [comment]
Why do you need to make a round trip to idsrv for signing in someone as a guest wouldn't it be easier to just set a local cookie for the guest in your app?

**Comment by rayph** [comment]
We have multiple apps

**Comment by leastprivilege** [comment]
I will look into it - in general i am not opposed to being able to pass custom data - but i don't think the `wfed` parameter is the right one. Looking at the spec - none of them seem right. So we might also just re-use `acr_values`.

**Comment by rayph** [comment]
Yes, none of the standard parameters feel quite right. The acr_values sounds interesting but I would be hoping for a convenient way to pass it through from the RP, right now I set WsFederationMessage.Wfed to the custom value and it gets forwarded/serialised as per the spec. Would I instead manually append "acr_values" query string parameter to the sign in redirect url? and would having a 'weird' value in SignInMessage.AcrValues cause issues elsewhere in IdSrv (or is it better to have a specific property for this)?

**Comment by leastprivilege** [comment]
done on dev. will push a 1.1 nuget soon

**Comment by rayph** [comment]
Very helpful. Thanks!
